lib/util/zip.ex

```diff
@@ -0,0 +1,100 @@
+# Copyright(c) 2015-2019 ACCESS CO., LTD. All rights reserved.
+
+use Croma
+
+defmodule Antikythera.Zip do
+  @moduledoc """
+  Wrapper module for `zip` command.
+
+  For the consistency in working with antikythera and other gears, scope of this module is limited under a temporary directory reserved by `Antikythera.Tmpdir.make/2`.
+
+  Functions only accept absolute paths for both source and resulting archive.
+  """
+
+  alias Croma.Result, as: R
+  alias Antikythera.Context
+  alias Antikythera.ExecutorPool.Id, as: EPoolId
+  alias AntikytheraCore.TmpdirTracker
+
+  @typep opts :: {:encryption, boolean} | {:password, String.t}
+
+  @doc """
+  Creates a ZIP file.
+
+  Encryption using a password is supported in which `encryption` option is set `true`.
+
+  ## Example
+    Tmpdir.make(context, fn tmpdir ->
+      src_path = tmpdir <> "/src.txt"
+      zip_path = tmpdir <> "/archive.zip"
+      File.write!(src_path, "text")
+      Antikythera.Zip.zip(context, zip_path, src_path, [encryption: true, password: "password"])
+      |> case do
+        {:ok, archive_path} ->
+          ...
+      end
+      ...
+    end)
+  """
+  defun zip(
+    context_or_epool_id :: v[EPoolId.t | Context.t],
+    zip_path            :: v[String.t],
+    src_path            :: v[String.t],
+    opts                :: v[list(opts)] \\ []
+  ) :: R.t(Path.t) do
+    epool_id = extract_epool_id(context_or_epool_id)
+    with(
+      {:ok, tmpdir} <- TmpdirTracker.get(epool_id),
+      :ok           <- validate_within_tmpdir(zip_path, tmpdir),
+      :ok           <- validate_within_tmpdir(src_path, tmpdir),
+      :ok           <- validate_path_exists(zip_path, false),
+      :ok           <- validate_path_exists(src_path, true),
```

Comment:
Probably I put toomuch emphasis on the case when `zip_path` is a *file* path...
Actually, it needs another way to treat with `zip_path` given as a *directory* path:
<table>
  <thead>
    <tr>
      <th>`zip_path` is a ...</th>
      <th>Result</th>
      <th>Conclution</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Existent file</td>
      <td>Error by `zip/3`</td>
      <td>Consistent</td>
    </tr>
      <td>Non-existent file</td>
      <td>OK by `zip/3`</td>
      <td>Consistent</td>
    </tr>
      <td>Existent directory</td>
      <td>OK by shell / Bad args for `zip/3`</td>
      <td>Inconsistent</td>
    </tr>
      <td>Non-existent directory</td>
      <td>Error by shell / Good args for `zip/3`</td>
      <td>Inconsistent</td>
    </tr>
  </tbody>
</table>

Reply:
Maybe it should accept *`Existent file`*; there is the *update* behavior in `zip` command.
Then, all it rejects is *`Non-existent directory`* as for `zip_path`.

Reply:
Actually, there is a way to crack the update behavior: update non-zip file by zip command.
But I wonder it can be validated in advance.
For now, it will be delegated to `zip` command, because it is too hard for a *gear* to look into detailed file format.

Reply:
I will further check edge cases in [`:zip.zip/3`](http://erlang.org/doc/man/zip.html#zip-3)...

Reply:
Maybe there are major changes:
- Behavior of `:zip.zip/3`:
  - when `zip_path` is a directory:
    - `mkdir -p` unless it exists
    - Returns `{:error, :eisdir}` even if the directory is created.
  - when `zip_path` is a file:
    - `mkdir -p` unless the directory exists
    - `touch` unless the file exists
    - Returns `{:error, :enoent}` unless `src_path` exists, even if the file is created.

Reply:
I decided to avoid some features which result in empty residues:
- ~when `zip_path` is a directory:~
  - ~`mkdir -p` unless it exists acf798e~
  - ~Returns `{:error, :eisdir}` even if the directory is created.~
  - => Invalidated; `zip_path` must be a file! 521977b 9f7681f
- when `zip_path` is a file:
  - `mkdir -p` unless the directory exists acf798e
  - `touch` unless the file exists 43f6c27
  - ~Returns `{:error, :enoent}` unless `src_path` exists, even if the file is created. 43f6c27~
  - => `touch` after checking `src_path` exists. There will not be an empty file. 
3b33d7a e17c625